### PR TITLE
OSD-15399 - Add field to specify cluster domain source

### PR DESCRIPTION
**Pull request by tnierman**
https://issues.redhat.com/browse/OSD-15399

---

These changes add an additional field, `DomainRef`, to specify the resource RMO should use as a source-of-truth for the cluster's domain. 

- `infra` indicates that the `.Status.APIServerURL` field of the `cluster` `infrastructure.config.openshift.io` object should be used. This behavior is the current default, and should remain largely unchanged.
- `hc` indicates that `.Spec.DNS.BaseDomain` of the `hostedcluster.hypershift.openshift.io` should be referenced instead. The `hostedcluster` is determined by analyzing the `hostedcontrolplane` object present in the namespace the `clusterurlmonitor` being reconciled.

### CODE REVIEW

File: config/crd/bases/_.yaml
Comment (dustman9000) [inline config/crd/bases/_.yaml:1]: Is this file needed?

Comment (tnierman) [reply config/crd/bases/_.yaml:1]: Interesting. I'd imagine not

File: controllers/clusterurlmonitor/clusterurlmonitor_supplement.go [hunk 1]
```diff
@@ -150,20 +157,75 @@ func (s *ClusterUrlMonitorReconciler) GetClusterUrlMonitor(req ctrl.Request) (v1
...
-func (s *ClusterUrlMonitorReconciler) GetClusterDomain() (string, error) {
+// GetClusterDomain returns the baseDomain for a cluster, using the correct method based on it's type
+func (s *ClusterUrlMonitorReconciler) GetClusterDomain(monitor v1alpha1.ClusterUrlMonitor) (string, error) {
+	if monitor.Spec.DomainRef == v1alpha1.ClusterDomainRefHCP {
+		return s.getHypershiftClusterDomain(monitor)
+	}
+	if monitor.Spec.DomainRef == v1alpha1.ClusterDomainRefInfra || monitor.Spec.DomainRef == "" {
+		return s.getInfraClusterDomain()
+	}
+	return "", fmt.Errorf("invalid .Spec.DomainRef detected: %s. Valid options are '%s' or '%s'", monitor.Spec.DomainRef, v1alpha1.ClusterDomainRefInfra, v1alpha1.ClusterDomainRefHCP)
```
Comment (dustman9000) [inline controllers/clusterurlmonitor/clusterurlmonitor_supplement.go:168]: See above for api validation

File: api/v1alpha1/clusterurlmonitor_types.go [hunk 2]
```diff
@@ -29,12 +29,25 @@ type ClusterUrlMonitorSpec struct {
...
 	// Foo is an example field of ClusterUrlMonitor. Edit ClusterUrlMonitor_types.go to remove/update
-	Prefix string  `json:"prefix,omitempty"`
-	Suffix string  `json:"suffix,omitempty"`
-	Port   string  `json:"port,omitempty"`
-	Slo    SloSpec `json:"slo,omitempty"`
+	Prefix    string           `json:"prefix,omitempty"`
+	Suffix    string           `json:"suffix,omitempty"`
+	Port      string           `json:"port,omitempty"`
+	Slo       SloSpec          `json:"slo,omitempty"`
+	DomainRef ClusterDomainRef `json:"domainRef,omitempty"`
```
[the remark this answers is not in the document]
Comment (tnierman) [reply api/v1alpha1/clusterurlmonitor_types.go:39]: TIL, thanks

**Comment by yithian** [comment]
seems sane to me

/lgtm

File: api/v1alpha1/clusterurlmonitor_types.go [hunk 3]
```diff
@@ -33,8 +33,24 @@ type ClusterUrlMonitorSpec struct {
...
+// ClusterDomainRef defines the object used determine the cluster's domain
+// By default, 'infra' is used, which references the 'infrastructures/cluster' object
+type ClusterDomainRef string
+
+var (
+	// ClusterDomainRefInfra indicates the clusterDomain should be determined from the 'infrastructures/cluster' object
+	ClusterDomainRefInfra ClusterDomainRef = "infra"
+
+	// ClusterDomainRefHC indicates the clusterDomain should be determined from the 'hostedCluster' object on the hypershift management cluster
+	ClusterDomainRefHC ClusterDomainRef = "hc"
```
Comment (yithian) [inline api/v1alpha1/clusterurlmonitor_types.go:51]:
```suggestion
	ClusterDomainRefHC ClusterDomainRef = "hcp"
```

**Review by sam-nguyen7 — COMMENTED** [review]
/lgtm
/approve
